2005-06-04  Gwenole Beauchesne  <gbeauchesne@mandriva.com>

	* Add direct jump support to x86-64.

--- qemu-0.7.0/exec-all.h.x86_64-direct-jump	2005-06-04 06:04:10.000000000 +0200
+++ qemu-0.7.0/exec-all.h	2005-06-04 06:59:35.000000000 +0200
@@ -153,6 +153,9 @@ int tlb_set_page(CPUState *env, target_u
 #if defined(__i386__) && !defined(_WIN32)
 #define USE_DIRECT_JUMP
 #endif
+#if defined(__x86_64__)
+#define USE_DIRECT_JUMP
+#endif
 
 typedef struct TranslationBlock {
     target_ulong pc;   /* simulated PC corresponding to this block (EIP + CS base) */
@@ -257,7 +260,7 @@ static inline void tb_set_jmp_target1(un
     asm volatile ("sync" : : : "memory");
     asm volatile ("isync" : : : "memory");
 }
-#elif defined(__i386__)
+#elif defined(__i386__) || defined(__x86_64__)
 static inline void tb_set_jmp_target1(unsigned long jmp_addr, unsigned long addr)
 {
     /* patch the branch destination */
@@ -333,7 +336,7 @@ do {\
 		  "1:\n");\
 } while (0)
 
-#elif defined(__i386__) && defined(USE_DIRECT_JUMP)
+#elif (defined(__i386__) || defined(__x86_64__)) && defined(USE_DIRECT_JUMP)
 
 /* we patch the jump instruction directly.  Use sti in place of the actual
    jmp instruction so that dyngen can patch in the correct result.  */
--- qemu-0.7.0/dyngen.c.x86_64-direct-jump	2005-06-04 06:04:10.000000000 +0200
+++ qemu-0.7.0/dyngen.c	2005-06-04 06:44:01.000000000 +0200
@@ -2549,6 +2549,17 @@ void gen_code(const char *name, host_ulo
                 if (rel->r_offset >= start_offset &&
 		    rel->r_offset < start_offset + copy_size) {
                     sym_name = strtab + symtab[ELFW(R_SYM)(rel->r_info)].st_name;
+                    if (strstart(sym_name, "__op_jmp", &p)) {
+                        int n;
+                        n = strtol(p, NULL, 10);
+                        /* __op_jmp relocations are done at
+                           runtime to do translated block
+                           chaining: the offset of the instruction
+                           needs to be stored */
+                        fprintf(outfile, "    jmp_offsets[%d] = %d + (gen_code_ptr - gen_code_buf);\n",
+                                n, rel->r_offset - start_offset);
+                        continue;
+                    }
                     get_reloc_expr(name, sizeof(name), sym_name);
                     type = ELF32_R_TYPE(rel->r_info);
                     addend = rel->r_addend;
